modules/ray_tracing/include/ray_tracing/TraceRayTools.h

```diff
@@ -0,0 +1,1052 @@
+                                                                        intersection_distance,
+                                                                        intersected_extrema,
+                                                                        hmax
+#ifdef DEBUG_RAY_INTERSECTIONS
+                                                                        ,
+                                                                        debug
+#endif
+                                                                        )
+                                                    : intersectQuad(start_point,
+                                                                    direction,
+                                                                    elem,
+                                                                    T::side_nodes_map[side][3],
+                                                                    T::side_nodes_map[side][2],
+                                                                    T::side_nodes_map[side][1],
+                                                                    T::side_nodes_map[side][0],
+                                                                    intersection_distance,
+                                                                    intersected_extrema,
+                                                                    hmax
+#ifdef DEBUG_RAY_INTERSECTIONS
+                                                                    ,
+                                                                    debug
+#endif
+                                                      );
+
+  if (intersected)
+    intersection_point = start_point + intersection_distance * direction;
+
+  return intersected;
+}
+
+/**
+ * @return Whether not the element is traceable
+ */
+bool isTraceableElem(const Elem * elem);
+/**
+ * @return Whether or not the element is traceable with adapativity
+ */
+bool isAdaptivityTraceableElem(const Elem * elem);
+
+/**
+ * Determines if a point is at a vertex of an element.
+ * @param elem The element
+ * @param point The point
+ * @return The local vertex ID the point is at, if any (invalid_point otherwise)
+ */
+unsigned short atVertex(const Elem * elem, const Point & point);
+
+/**
+ * Determines if a point is at a vertex on the side of en element.
+ * @param elem The element
+ * @param point The point
+ * @param side The side
+ * @return The local vertex ID the point is at, if any (invalid_point otherwise)
+ */
+unsigned short atVertexOnSide(const Elem * elem, const Point & point, const unsigned short side);
+
+/**
+ * Returns the number of nodes on a side for an Elem that is not a Pyramid or Prism.
+ */
+template <typename T>
+inline typename std::enable_if<!std::is_base_of<Pyramid, T>::value &&
+                                   !std::is_base_of<Prism, T>::value,
+                               unsigned short>::type
+nodesPerSide(const unsigned short)
```

Comment:
Yes. But we need SFINAE here because typically we only have an `Elem` object. In order to access the low level structures which belong to the derived elements, i.e., `Pyramid5::nodes_per_side`, we need to keep track of what derived class we have. Which we do in tracing. These checks allow for compile-time access.